### PR TITLE
Call and operator write ergonomics

### PR DESCRIPTION
**Pull request by kddnewton**
Fixes https://github.com/ruby/yarp/issues/1236

This is definitely a tradeoff, because now we don't have a single call node anymore. However, I think this is acceptable considering the call node couldn't represent that actually 2 methods get called when you have these nodes. So it would have been confusing to client libraries anyway.

### CODE REVIEW

**Review by jemmaissroff — APPROVED** [review]
I like this change, and like that it mirrors the structure we already have in place here

File: config.yml [hunk 1]
```diff
@@ -625,50 +656,70 @@ nodes:
...
+        type: location?
+      - name: flags
+        type: flags
+        kind: CallNodeFlags
+      - name: read_name
+        type: string
+      - name: write_name
+        type: string
+      - name: operator
+        type: constant
```
Comment (eregon) [inline config.yml:682]: Duplicating the many fields of CallNode here feels rather like a code smell or anti-pattern.
And e.g. the logic to deal with visibility, arguments, block and various other aspects of calls now need to be duplicated for these 4 nodes, or all of these non-location parameters need to be passed to a common helper function.
Something to help with those concerns would be to track visibility in CallNodeFlags, and #1310. Then those operators calls are probably fairly simple to compile with little duplication.
I think before it would simply be enough to replace the receiver by a temporary variable and be done with it + a copy of the CallNode with/without the `=`, although we haven't written the code for that yet in TruffleRuby, so it might be slightly more complicated.

FWIW here is how the parser gem represents it:
```
$ CODE="foo.bar += baz" && ruby -rparser/current -e 'pp Parser::CurrentRuby.parse(ARGV[0])' "$CODE"     
s(:op_asgn,
  s(:send,
    s(:send, nil, :foo), :bar), :+,
  s(:send, nil, :baz))
```
Which is quite similar to current YARP/before this PR.

And this is how CRuby's current parser represents it:
```
$ ruby --dump=parsetree -e 'foo.bar += baz'                  
# @ NODE_SCOPE (id: 4, line: 1, location: (1,0)-(1,14))
# +- nd_tbl: (empty)
# +- nd_args:
# |   (null node)
# +- nd_body:
#     @ NODE_OP_ASGN2 (id: 3, line: 1, location: (1,0)-(1,14))*
#     +- nd_recv:
#     |   @ NODE_VCALL (id: 0, line: 1, location: (1,0)-(1,3))
#     |   +- nd_mid: :foo
#     +- nd_next->nd_vid: :bar
#     +- nd_next->nd_mid: :+
#     +- nd_value:
#         @ NODE_VCALL (id: 1, line: 1, location: (1,11)-(1,14))
#         +- nd_mid: :baz
```
Which seems closer to this change.

@enebo Does this change help?
What's the difference for https://gist.github.com/enebo/13cc0fcba7159b82a2be60828dd45e66?
What would `buildCallOperatorWrite()` look like with this change?

In summary I'm not really enthusiastic about this change but also we discussed it and I think it's fine to merge.
The addition of both `read_name` and `write_name` is certainly nice.

Comment (enebo) [reply config.yml:682]: @eregon The main points for why I think it is fine:

  1. receiver/target is now no longer nested into another node
  2. The original node was not usable as it existed for read or write
  3. the names for read/write are not things we have to process (e.g chomp write to get read). we just use them (and this also be related to our symbol discussion in other bug since these should eventually be in constant pool).
  4. This is smaller due to one less node

My take on needing common methods is that I already need common methods because I will still be supporting the existing parser for years/ever (pure-Java mode).  I already have one aspect of that already.

So this is a big step forward at the cost of some repeated fields.

Comment (eregon) [reply config.yml:682]: Right, those are good points, thanks for explaining them.
1. makes perfect sense to me (could also be independently but might be weird then).
2. indeed not usable as-is (and no copy/with method yet in Java but we could add some if useful)
3. agreed
